### PR TITLE
TLDEquality refactored to perform an in-line byte comparison

### PR DESCRIPTION
**Pull request by apmoriarty**
TLDEquality refactored to perform an in-line byte comparison instead of delegating to expensive TLD utility methods

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/function/TLDEquality.java [hunk 1]
```diff
@@ -23,8 +21,20 @@ public class TLDEquality implements Equality {
...
+        ByteSequence otherCf = other.getColumnFamilyData();
+
+        int dotCount = 0;
+        int len = Math.min(keyCf.length(), otherCf.length());
+        for (int i = 0; i < len; i++) {
+            byte a = keyCf.byteAt(i);
+            byte b = otherCf.byteAt(i);
+            if (a != b) {
+                return false;
+            } else if (a == '.' && ++dotCount == 3) {
```
Comment (MiguelRicardos) [inline warehouse/query-core/src/main/java/datawave/query/function/TLDEquality.java:34]: why 3?

**Review by MiguelRicardos — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/function/TLDEquality.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/function/TLDEquality.java:34]: A root uid only has two dots, a child uid has three or more. Thus, if a third dot is encountered and everything up to that point has matched, then the root uids are the same.

File: warehouse/query-core/src/main/java/datawave/query/function/TLDEquality.java [hunk 2]
```diff
@@ -23,8 +21,20 @@ public class TLDEquality implements Equality {
...
+                return true;
+            }
+        }
+        return true;
```
Comment (d-hwang) [inline warehouse/query-core/src/main/java/datawave/query/function/TLDEquality.java]: maybe this doesn't happen, but this default return will make any comparison to an empty string true.

Comment (d-hwang) [reply warehouse/query-core/src/main/java/datawave/query/function/TLDEquality.java]: return len == 0 ? false : true;

??

Comment (d-hwang) [reply warehouse/query-core/src/main/java/datawave/query/function/TLDEquality.java]: but then this will make empty value comparisons to false..

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/function/TLDEquality.java]: I like that solution. The return statement could be simplified to `return len != 0;`

Comment (d-hwang) [reply warehouse/query-core/src/main/java/datawave/query/function/TLDEquality.java]: existing code will also evaluate these two to be the same? a.b.c and a.b.c1
maybe each part of UID has fixed length, but comparing values up to shorter length sounds a bit dangerous to me.

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/function/TLDEquality.java]: Given the structure of a uid I don't think that case is possible, but I added a test to document that behavior anyway.

**Review by d-hwang — APPROVED** [review]

**Review by MiguelRicardos — APPROVED** [review]

**Pull request merged**